ext/dom/token_list.c

```diff
@@ -0,0 +1,729 @@
+	dom_token_list_object *intern = TOKEN_LIST_GET_INTERNAL();
+	dom_token_list_ensure_set_up_to_date(intern);
+	HashTable *token_set = TOKEN_LIST_GET_SET(intern);
+	for (uint32_t i = 0; i < argc; i++) {
+		zend_hash_del(token_set, Z_STR(args[i]));
+	}
+
+	/* 3. Run the update steps. */
+	dom_token_list_update(intern);
+}
+
+/* https://dom.spec.whatwg.org/#dom-domtokenlist-toggle */
+PHP_METHOD(Dom_TokenList, toggle)
+{
+	zend_string *token;
+	bool force, force_not_given = true;
+	ZEND_PARSE_PARAMETERS_START(1, 2)
+		Z_PARAM_PATH_STR(token)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_BOOL_OR_NULL(force, force_not_given)
+	ZEND_PARSE_PARAMETERS_END();
+
+	/* Steps 1 - 2 */
+	if (!dom_validate_token(token)) {
+		RETURN_THROWS();
+	}
+
+	/* 3. If this’s token set[token] exists, then: */
+	dom_token_list_object *intern = TOKEN_LIST_GET_INTERNAL();
+	dom_token_list_ensure_set_up_to_date(intern);
+	HashTable *token_set = TOKEN_LIST_GET_SET(intern);
+	zval *found_token = zend_hash_find(token_set, token);
+	if (found_token != NULL) {
+		ZEND_ASSERT(XtOffsetOf(Bucket, val) == 0 && "the cast only works if this is true");
+		Bucket *bucket = (Bucket *) found_token;
+
+		/* 3.1. If force is either not given or is false, then remove token from this’s token set,
+		 *      run the update steps and return false. */
+		if (force_not_given || !force) {
+			zend_hash_del_bucket(token_set, bucket);
+			dom_token_list_update(intern);
+			RETURN_FALSE;
+		}
+
+		/* 3.2. Return true. */
+		RETURN_TRUE;
+	}
+	/* 4. Otherwise, if force not given or is true, append token to this’s token set,
+	 *    run the update steps, and return true. */
+	else if (force_not_given || force) {
+		dom_add_token(token_set, token);
+		dom_token_list_update(intern);
+		RETURN_TRUE;
+	}
+
+	/* 5. Return false. */
+	RETURN_FALSE;
+}
+
+/* https://dom.spec.whatwg.org/#dom-domtokenlist-replace */
+PHP_METHOD(Dom_TokenList, replace)
+{
+	zend_string *token, *new_token;
+	ZEND_PARSE_PARAMETERS_START(2, 2)
+		Z_PARAM_PATH_STR(token)
+		Z_PARAM_PATH_STR(new_token)
+	ZEND_PARSE_PARAMETERS_END();
+
+	/* Steps 1 - 2 */
+	if (!dom_validate_token(token) || !dom_validate_token(new_token)) {
+		RETURN_THROWS();
+	}
+
+	/* 3. If this’s token set does not contain token, then return false. */
+	dom_token_list_object *intern = TOKEN_LIST_GET_INTERNAL();
+	dom_token_list_ensure_set_up_to_date(intern);
+	HashTable *token_set = TOKEN_LIST_GET_SET(intern);
+	zval *found_token = zend_hash_find(token_set, token);
+	if (found_token == NULL) {
+		RETURN_FALSE;
+	}
+
+	/* 4. Replace token in this’s token set with newToken. */
+	ZEND_ASSERT(XtOffsetOf(Bucket, val) == 0 && "the cast only works if this is true");
+	Bucket *bucket = (Bucket *) found_token;
+	if (zend_hash_set_bucket_key(token_set, bucket, new_token) == NULL) {
+		/* It already exists, remove token instead. */
+		zend_hash_del_bucket(token_set, bucket);
+	} else {
+		Z_STR(bucket->val) = new_token;
+	}
```

Comment:
I will be honest here, I don't understand/forgot how our buckets work, so this is a bit magical to me rn.

Reply:
It just uses the fact that value is at the start of the bucket and then uses the bucket API to rename the key.
If the key already exists then it removes the entry instead because they keys are a set, so no duplicate entries are allowed.
In the else branch I just overwrite the string to point to the new token. I made the value and key both equal to the same string for convenience.

Reply:
ACK